docs/design-documents/20240812-recover-from-pipeline-errors.md

```diff
@@ -0,0 +1,194 @@
+# Recover from Pipeline Errors
+
+## Introduction
+
+### Goals
+
+The primary goal of this design document is to introduce a robust error-handling
+mechanism within Conduit to ensure the reliability and resilience of data
+pipelines. The objectives include:
+
+- **Error Classification:** Distinguish between transient (recoverable) and
+  fatal (non-recoverable) errors.
+- **Automatic Recovery:** Implement a mechanism to automatically restart
+  pipelines in the event of transient errors with configurable backoff
+  strategies.
+- **Pipeline State Management:** Introduce new pipeline states to clearly
+  indicate the status of recovery processes.
+- **Configurable Backoff Strategy:** Allow users to configure backoff parameters
+  to suit different production environments.
+
+## Background
+
+Conduit's engine is currently very strict, stopping the pipeline and putting it
+into a degraded state whenever an error occurs. This approach is not optimal for
+production environments where transient errors, such as connectivity issues, may
+cause frequent disruptions. To enhance reliability, Conduit should include a
+mechanism to automatically restart failing pipelines, leveraging the fact that
+Conduit already tracks acknowledged record positions, making restarts a safe
+operation with no risk of data loss (assuming connectors are implemented
+correctly).
+
+## Error classification
+
+### Fatal errors
+
+Fatal errors are defined as non-recoverable issues that require manual
+intervention. When a fatal error occurs, the pipeline will be stopped
+permanently, and no further attempts will be made to restart it. This approach
+ensures that critical issues are addressed promptly and prevents Conduit from
+entering an endless loop of restarts.
+
+### Transient errors
+
+Transient errors are recoverable issues that do not require manual intervention.
+Examples include temporary network failures or timeouts. For such errors,
+Conduit will automatically attempt to restart the pipeline using a backoff delay
+strategy.
+
+- **Default Behavior:** By default, all errors are considered transient and
+  recoverable. A connector developer must explicitly mark an error as fatal if
+  it should cause the pipeline to stop permanently.
+
+## Recovery mechanism
+
+### Backoff delay strategy
+
+To manage transient errors and prevent constant pipeline restarts, Conduit will
+implement a backoff delay strategy. This strategy introduces a waiting period
+before attempting to restart the pipeline, allowing time for issues to resolve
+themselves. The backoff parameters will be configurable to allow customization
+based on specific production needs.
+
+- **Configurable Parameters:**
+  - **Minimum Delay Before Restart:** Default: 1 second
+  - **Maximum Delay Before Restart:** Default: 1 minute
+  - **Backoff Factor:** Default: 2
+  - **Maximum Number of Retries:** Default: Infinite
```

Comment:
IMHO, we should limit the number of retries. If, after many retries, an error is still not gone, then it's very likely not a transient error (which the pipeline engine thinks it is).

Reply:
In case of a fatal error, you're completely right, we will be trying again and again forever without the possibility to recover, which is not ideal. But let's say we do limit the maximum number of retries - what would be a sane default for this limit?

This number essentially defines the time window in which the issue should be resolved, otherwise the pipeline will be degraded and an operator needs to step in to resume the pipeline. If I set up a pipeline in production, I probably want it to run forever without my intervention (or at least with as little intervention as possible). Wouldn't that speak for retrying every now and then?

How about we increase the maximum delay to something like 10 minutes instead? 🤔

Reply:
We discussed this in an "offline" meeting, the default can stay `infinite`, as this will be configurable any way and called out in the docs.